### PR TITLE
[Incremental] Supports per-type-body fingerprints, off-by-default

### PR DESCRIPTION
**Pull request by davidungar**
1. Changed granularity of (fine-grained) dependency graph traversal to support tracking changes to individual node fingerprints.
2. Changed semantics of transitive traversal to simplify meaning of a mark from "was a dependent of a cascading job" to "this node has been traversed".
3. Adds `-enable-type-fingerprints` disabled by default.
4. Changes driver handling of jobs when one job fails. Don't bother marking all unfinished jobs as cascading if they were previously thought to be. Just mark then as non-cascading-build required. On next run, if they are cascading the dependent jobs will get run.
5. When the option above is set, the parser will hash tokens within a nominal type body or extension body into a per-body hash instead of the whole-file interface hash.
6. Adjust tests accordingly.

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please smoke test

File: include/swift/AST/DeclContext.h [hunk 1]
```diff
@@ -802,6 +802,12 @@ class IterableDeclContext {
   // Some Decls are IterableDeclContexts, but not all.
   static bool classof(const Decl *D);
 
+  /// Return a hash of all tokens in the body for dependency analysis.
+  /// Empty string means no hash available.
+  StringRef getBodyFingerprint() const;
```
Comment (CodaFi) [inline include/swift/AST/DeclContext.h]: I'm not a fan of overloading the meaning of the empty string in APIs like this.  I think this should return an `Optional<StringRef>` and clients should assert it's not ever the empty string.

**Review by CodaFi — APPROVED** [review]
Can't speak to the driver changes, but the layering changes and the test updates are pretty exciting.

**Comment by davidungar** [comment]
@swift-ci please smoke test os x

File: include/swift/AST/DeclContext.h (hunk 1 above)
Comment (davidungar) [reply include/swift/AST/DeclContext.h]: Nice catch! I've changed it to an optional, and also changed the \c StringRef to a \c std::string in order to make the memory safety less brittle in the face of request evaluator evolution.

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
@swift-ci please smoke test linux platform

**Comment by CodaFi** [comment]
@swift-ci test Windows platform

**Comment by CodaFi** [comment]
@swift-ci test Windows platform

**Pull request merged**